metrics.nim

```diff
@@ -759,59 +806,72 @@ when defined(metrics) and defined(linux):
-        value: selfStat[19].parseFloat() / ticks + btime,
-        timestamp: timestamp,
-      ),
-      Metric(
-        name: "process_cpu_seconds_total", # Total user and system CPU time spent in seconds.
-        value: (selfStat[11].parseFloat() + selfStat[12].parseFloat()) / ticks,
-        timestamp: timestamp,
-      ),
-    ]
+    try:
+      if btime == 0:
+        # we couldn't access /proc
+        return
+
+      # the content of /proc/self/stat looks like this (the command name may contain spaces):
+      #
+      # $ cat /proc/self/stat
+      # 30494 (cat) R 3022 30494 3022 34830 30494 4210688 98 0 0 0 0 0 0 0 20 0 1 0 73800491 10379264 189 18446744073709551615 94060049248256 94060049282149 140735229395104 0 0 0 0 0 0 0 0 0 17 6 0 0 0 0 0 94060049300560 94060049302112 94060076990464 140735229397011 140735229397031 140735229397031 140735229403119 0
+      let selfStat = readFile("/proc/self/stat").split(") ")[^1].split(' ')
+      result[@[]] = @[
+        Metric(
+          name: "process_virtual_memory_bytes", # Virtual memory size in bytes.
+          value: selfStat[20].parseFloat(),
+          timestamp: timestamp,
+        ),
+        Metric(
+          name: "process_resident_memory_bytes", # Resident memory size in bytes.
+          value: selfStat[21].parseFloat() * pagesize,
+          timestamp: timestamp,
+        ),
+        Metric(
+          name: "process_start_time_seconds", # Start time of the process since unix epoch in seconds.
+          value: selfStat[19].parseFloat() / ticks + btime,
+          timestamp: timestamp,
+        ),
+        Metric(
+          name: "process_cpu_seconds_total", # Total user and system CPU time spent in seconds.
+          value: (selfStat[11].parseFloat() + selfStat[12].parseFloat()) / ticks,
+          timestamp: timestamp,
+        ),
+      ]
+
+      for line in lines("/proc/self/limits"):
+        if line.startsWith("Max open files"):
+          result[@[]].add(
+            Metric(
+              name: "process_max_fds", # Maximum number of open file descriptors.
+              value: line.split(whitespaceRegex)[3].parseFloat(), # a simple `split()` does not combine adjacent whitespace
+              timestamp: timestamp,
+            )
+          )
+          break
 
-    for line in lines("/proc/self/limits"):
-      if line.startsWith("Max open files"):
+      result[@[]].add(
+        Metric(
+          name: "process_open_fds", # Number of open file descriptors.
+          value: toSeq(walkDir("/proc/self/fd")).len.float64,
+          timestamp: timestamp,
+        )
+      )
+    except CatchableError:
+      try:
         result[@[]].add(
           Metric(
-            name: "process_max_fds", # Maximum number of open file descriptors.
-            value: line.split(whitespaceRegex)[3].parseFloat(), # a simple `split()` does not combine adjacent whitespace
+            name: "metric_error",
```

Comment:
The best way to hide an error I've seen so far. Nobody will add this metric name to Grafana, so no one will see it.

Reply:
you're doing the same thing when `/proc` is not available, except without the metric - what's your suggestion? discard? crash application if `/proc` has issues?

Reply:
"/proc" not being mounted is not a showstopper, nor is it something the API user needs to be informed about by this library.